mpqp/core/instruction/gates/controlled_gate.py

```diff
@@ -39,3 +41,131 @@ def __init__(
+        control, target = self.controls[0], self.targets[0]
+        min_ = min(control, target)
+        control -= min_
+        target -= min_
+
+        nb_qubit = abs(control - target) + 1
+        zero = np.diag([1, 0])
+        one = np.diag([0, 1])
+        I2 = np.eye(2)
+
+        control_matrix = zero if control == 0 else I2
+        target_matrix = (
+            one
+            if control == 0
+            else (self.non_controlled_gate.to_matrix() if target == 0 else I2)
+        )
+
+        for i in range(1, nb_qubit):
+            if i == control:
+                target_matrix = np.kron(target_matrix, one)
+            elif i == target:
+                target_matrix = np.kron(
+                    target_matrix, self.non_controlled_gate.to_matrix()
+                )
+            else:
+                target_matrix = np.kron(target_matrix, I2)
+
+        for i in range(1, nb_qubit):
+            if i == control:
+                control_matrix = np.kron(control_matrix, zero)
+            else:
+                control_matrix = np.kron(control_matrix, I2)
+
+        return control_matrix + target_matrix  # pyright: ignore[reportReturnType]
+
+    def _multi_control_gate_to_matrix(self) -> Matrix:
+        import math
+
+        import numpy as np
+
+        from mpqp.core.instruction.gates.native_gates import SWAP
+
+        I2 = np.eye(2)
+
+        min_qubit = min(min(self.controls), min(self.targets))
+        max_qubit = max(max(self.controls), max(self.targets))
+        nb_qubits = max_qubit - min_qubit + 1
+
+        canonical_matrix = self.to_canonical_matrix()
+        while canonical_matrix.shape[0] < 2**nb_qubits:
+            canonical_matrix = np.kron(canonical_matrix, I2)
+
+        matrix = np.eye(2**nb_qubits, dtype=np.complex64)
+        qubit_types = {i: "None" for i in range(nb_qubits)}
+
+        for i, _ in enumerate(self.controls):
+            qubit_types[i] = "control"
+        for i, _ in enumerate(self.targets, start=len(self.controls)):
+            qubit_types[i] = "target"
+
+        def swap_and_update(matrix: Matrix, idx_a: int, idx_b: int):
+            swap_matrix = SWAP(idx_a, idx_b).to_matrix()
+            extended_swap_matrix = np.eye(2**nb_qubits)
+            start = min(idx_b, idx_a)
+            if start == 0:
+                extended_swap_matrix = swap_matrix
+            if start != 0:
+                extended_swap_matrix = I2
+                for _ in range(1, start - 1):
+                    extended_swap_matrix = np.kron(extended_swap_matrix, I2)
+                extended_swap_matrix = np.kron(extended_swap_matrix, swap_matrix)
+            for _ in range(int(math.log2(extended_swap_matrix.shape[0])), nb_qubits):
+                extended_swap_matrix = np.kron(extended_swap_matrix, I2)
+            return np.dot(matrix, extended_swap_matrix)
+
+        for control in sorted(self.controls):
+            control -= min_qubit
+            if qubit_types[control] != "control":
```

Comment:
you're iterating over the controls, why wouldn't it be a control

Reply:
its the case on multiple control